crates/bitwarden-crypto/src/sensitive/string.rs

```diff
@@ -0,0 +1,241 @@
+use std::borrow::Cow;
+
+use schemars::JsonSchema;
+use zeroize::{Zeroize, Zeroizing};
+
+use crate::{Sensitive, SensitiveVec};
+
+/// A sensitive string that supports string operations.
+///
+/// Important: The `SensitiveString` protects against reallocations in the internal string. However,
+/// be careful when using any str or byte slices as taking ownership of them will create a copy
+/// which is not protected.
+#[derive(Eq, Clone, Zeroize)]
+pub struct SensitiveString {
+    inner: Zeroizing<String>,
+}
+
+impl std::fmt::Debug for SensitiveString {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Sensitive")
+            .field("type", &std::any::type_name::<String>())
+            .field("value", &"********")
+            .finish()
+    }
+}
+
+impl SensitiveString {
+    #[inline(always)]
+    pub fn new(inner: String) -> Self {
+        Self {
+            inner: Zeroizing::new(inner),
+        }
+    }
+
+    /// Creates a new empty `SensitiveString` with at least the specified capacity.
+    pub fn with_capacity(len: usize) -> Self {
+        Self {
+            inner: Zeroizing::new(String::with_capacity(len)),
+        }
+    }
+
+    /// Extend the size of the string by `size`.
+    ///
+    /// Internally creates a new string with the new size and copies the old string into it.
+    pub fn extend_size(&mut self, size: usize) {
+        let mut new_inner = Zeroizing::new(String::with_capacity(size));
+        new_inner.push_str(&self.inner);
+
+        self.inner = new_inner;
+    }
+
+    /// Extends the capacity of the string to `size` if it is larger than the current capacity.
+    pub fn extend_spec(&mut self, size: usize) {
+        if size > self.inner.capacity() {
+            self.extend_size(size);
+        }
+    }
+
+    /// Appends a given char onto the end of this `BitString`
+    ///
+    /// If the capacity of the `BitString` is not large enough it will zeroize the current string
+    /// and create a new string with the new size.
+    pub fn push(&mut self, c: char) {
+        self.extend_spec(self.inner.len() + c.len_utf8());
+        self.inner.push(c);
+    }
+
+    /// Appends a given string slice onto the end of this `BitString`
+    ///
+    /// If the capacity of the `BitString` is not large enough it will zeroize the current string
+    /// and create a new string with the new size.
+    pub fn push_str(&mut self, s: &str) {
+        self.extend_spec(self.inner.len() + s.len());
+        self.inner.push_str(s);
+    }
+
+    /// Expose the inner value as a str slice.
+    ///
+    /// Warning: Taking ownership of the inner value will create a copy of the string.
+    pub fn as_str(&self) -> &str {
+        &self.inner
+    }
+
+    /// Expose the inner value as a byte slice.
+    ///
+    /// Warning: Taking ownership of the inner value will create a copy of the string.
+    pub fn as_bytes(&self) -> &[u8] {
```

Comment:
Yes, unfortunately we do need direct access to the str and bytes in many scenarios. But as you said we need to be careful with how they are used since taking ownership of them means you need to care about zeroing it out.

I was leaning towards moving them into a separate trait so that we have more fine grained control of the usage, but I think ultimately we will have to rely on reviewers to be vigilant when using that trait.